Source/WebCore/html/CanvasBase.h

```diff
@@ -90,6 +90,8 @@ class CanvasBase {
     void setOriginTainted() { m_originClean = false; }
     bool originClean() const { return m_originClean; }
 
+    void setHadGradientFill() { m_hadGradientFill = true; }
```

Comment:
Adding custom ad-hoc state to the CanvasBase will result in hard to maintain Canvas implementation down the line.
Adding ad-hoc state to the algorithm creates a problem where you need to run security analysis on each possible state.
Would it be possible to implement one, stateless algorithm that would be good?

If the gradient algorithm is good enough for injecting noise for the gradient case, could it be used also in non-gradient case? This would fix the corner-cases where a gradient would be drawn to a canvas a, canvas a to canvas b, and canvas b would be read back, injecting the noise in b read.